---
title: "GATA hexamers"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library("factoextra")
library("ade4")
library(dplyr)
```

## Background

The transcription factor ELT-2 is known to bind the sequence TGATAA with highest affinity. However, other variants of the NGATAN motif may contribute to the regulatory logic recognized by ELT-2.

Here is an analysis of roughly 10,000 DNA sequences that ELT-2 binds in any of three developmental stages, as determined by ChIP-seq.

```{r show data}
ngatan.by.peak = readRDS("ngatan.by.peak.rds") # a "peak" is the entity identified by ChIP-seq analysis.  A peak is d
dim(ngatan.by.peak)
mappings = as.integer(unlist(str_split(rownames(ngatan.by.peak),":"))[(1:10185)*3]) + 1

# Contingency tables on # of peaks with X number of givenmotif
newtab = table(ngatan.by.peak[,'TGATAA'],mcols(ap[mappings])$k4cluster)[1:4,]
nPeaks_w_nTGATAA = rbind(newtab, colSums(table(ngatan.by.peak[,'TGATAA'], mcols(ap[mappings])$k4cluster)[5:11,]))
rownames(nPeaks_w_nTGATAA) <- c("0","1","2","3","≥4")
mosaicplot(nPeaks_w_nTGATAA, shade=T,
           main="TGATAA occurance among peaks\nbroken down by developmental class",
           xlab="peaks with specified # of TGATAAs",
           ylab="developmental class")

nPeaks_w_nTGATAT=rbind(table(ngatan.by.peak[,'TGATAT'],mcols(ap[mappings])$k4cluster)[1:3,], colSums(table(ngatan.by.peak[,'TGATAT'],mcols(ap[mappings])$k4cluster)[4:10,]))
rownames(nPeaks_w_nTGATAT) <- c("0","1","2","≥3")
mosaicplot(nPeaks_w_nTGATAT, shade=T,
           main="TGATAT occurance among peaks\nbroken down by developmental class",
           xlab="peaks with specified # of TGATATs",
           ylab="developmental class")

nPeaks_w_nAGATAA=rbind(table(ngatan.by.peak[,'AGATAA'],mcols(ap[mappings])$k4cluster)[1:3,], colSums(table(ngatan.by.peak[,'AGATAA'],mcols(ap[mappings])$k4cluster)[4:8,]))
rownames(nPeaks_w_nAGATAA) <- c("0","1","2","≥3")
mosaicplot(nPeaks_w_nAGATAA, shade=T,
           main="AGATAA occurance among peaks\nbroken down by developmental class",
           xlab="peaks with specified # of AGATAAs",
           ylab="developmental class")

nPeaks_w_nAGATAT=rbind(table(ngatan.by.peak[,'AGATAT'],mcols(ap[mappings])$k4cluster)[1:3,], colSums(table(ngatan.by.peak[,'AGATAT'],mcols(ap[mappings])$k4cluster)[4:8,]))
rownames(nPeaks_w_nAGATAT) <- c("0","1","2","≥3")
mosaicplot(nPeaks_w_nAGATAT, shade=T,
           main="AGATAT occurance among peaks\nbroken down by developmental class",
           xlab="peaks with specified # of AGATATs",
           ylab="developmental class")

nPeaks_w_nAGATAG=rbind(table(ngatan.by.peak[,'AGATAG'],mcols(ap[mappings])$k4cluster)[1:3,], colSums(table(ngatan.by.peak[,'AGATAG'],mcols(ap[mappings])$k4cluster)[4:6,]))
rownames(nPeaks_w_nAGATAG) <- c("0","1","2","≥3")
mosaicplot(nPeaks_w_nAGATAG, shade=T,
           main="AGATAG occurance among peaks\nbroken down by developmental class",
           xlab="peaks with specified # of AGATAGs",
           ylab="developmental class")


# Do the non-TGATAA regions have greater/less than expected usage
# of secondary motifs?
zeroTGATAA.ix = ngatan.by.peak[,'TGATAA'] == 0 

# 0 TGATAA; breakdown of TGATAT (need to aggregrate counts ≥ 3)
submotif = 'TGATAT' 
TGATAT_zero_TGATAA = table(ngatan.by.peak[zeroTGATAA.ix, 'TGATAT'],
                           mcols(ap[mappings][zeroTGATAA.ix])$k4cluster)
TGATAT_zero_TGATAA = rbind(TGATAT_zero_TGATAA[1:2,],colSums(TGATAT_zero_TGATAA[3:7,]))
rownames(TGATAT_zero_TGATAA) <- c("0","1","≥2")

mosaicplot(TGATAT_zero_TGATAA,
           shade=T,
           main="TGATAT occurance among peaks\nthat have 0 TGATAAs",
           sub="Larval-high has slightly more peaks with 2 TGATATs than expected\nand fewer peaks with no TGATATs than expected.\nL3-high has more peaks with 0 TGATAAs than expected.",
           ylab="kcluster")

# 0 TGATAA; breakdown of AGATAA (need to aggregrate counts ≥ 3)
AGATAA_zero_TGATAA = table(ngatan.by.peak[zeroTGATAA.ix, 'AGATAA'],
                           mcols(ap[mappings][zeroTGATAA.ix])$k4cluster)
AGATAA_zero_TGATAA = rbind(AGATAA_zero_TGATAA[1:3,],colSums(AGATAA_zero_TGATAA[4:6,]))
rownames(AGATAA_zero_TGATAA) <- c("0","1","2","≥3")
mosaicplot(AGATAA_zero_TGATAA, 
           shade=T,
           main="AGATAA occurance among peaks\nthat have 0 TGATAAs",
           xlab="peaks with specified # of AGATAAs",
           sub="AGATAA occurance in TGATAA-0 peaks is not different between dev. classes",
           ylab="developmental class")
ngatan.by.peak[1:3,]
ngatan.tbl = as.data.frame(ngatan.by.peak)
colnames(ngatan.tbl) <- c('region','hexamer','count')
totals = ngatan.tbl %>% group_by(hexamer) %>% summarize(total.hex = sum(count)) 
ggplot(totals, aes(x=reorder(hexamer,-total.hex),y=total.hex)) + geom_col() +theme(axis.text.x = element_text(angle = 90))
```
As expected, TGATAA is the most common motif overall. Others, especially AGATAA, have been studied as secondary targets.


How does the occurance of GATA variants compare with those counted in randomly permuted regions?

```{r permuted regions}
permuted_NGATAN.tab = readRDS("permuted_NGATAN.tab.rds")
permuted_10.counts = colSums(permuted_NGATAN.tab)
ave_hexamer_count = round(sum(permuted_10.counts)/10)
ELT2_binding_counts = colSums(ngatan.by.peak)
# scale the permuted counts to the average hexamer count size
permuted_10.prop = permuted_10.counts/sum(permuted_10.counts)
permuted_10.sim = round(permuted_10.prop * ave_hexamer_count)
ELT2_hexamers_vs_random_regions = rbind(ELT2_binding_counts, permuted_10.sim)
mosaicplot(ELT2_hexamers_vs_random_regions, shade=T, las=1, type="pearson", cex.axis=.7)
```

### PCA analysis: Contribution of secondary GATA variants
```{r thresholding}
variance.by.peak = apply(ngatan.by.peak, 1, var)
for (upper_quantile in c(.25,.05,.01))
{
  upper.variance.i = variance.by.peak >= quantile(1-upper_quantile, type=1)
  ngatan.upper.tbl = as.data.frame(ngatan.by.peak[upper.variance.i,])
  colnames(ngatan.upper.tbl) <- c('region','hexamer','count')
  totals.upper = ngatan.upper.tbl %>% group_by(hexamer) %>% summarize(total.hex = sum(count))
  plot<- ggplot(totals.upper, aes(x=reorder(hexamer,-total.hex),y=total.hex)) + geom_col() +theme(axis.text.x = element_text(angle = 90)) + ggtitle(sprintf("top %.0f%% by variance", upper_quantile*100))
  print(plot)
}
```
Variance thresholds do not appear do select for changes in relative occurance of the motifs.

---




```{r PCA}
ngatan.pca = princomp(ngatan.by.peak)
# The ade4 PCA implementation returns an error
# > ngatan.dudi.pca = dudi.pca(ngatan.by.peak)
# Error in v * row.w : non-numeric argument to binary operator

fviz_pca_biplot(ngatan.pca, label="var") + ggtitle("")

fviz_pca_biplot(ngatan.pca, axes=c(1,3), label="var") + ggtitle("")

```
The data appear in strips, probably because it is count data.

Although much of the variance is explained by TGATAA, AGATAA and TGATAT appear to contribute orthogonally to TGATAA, and to each other.

Let's look at this again by considering everything but TGATAA.
```{r PCA_nonTGATAA}
nonTGATAA.pca = princomp(ngatan.by.peak[,base::grep("TGATAA", colnames(ngatan.by.peak),invert=T)])

fviz_pca_biplot(nonTGATAA.pca, label="var") + ggtitle("TGATAA excluded")

fviz_pca_biplot(nonTGATAA.pca, axes=c(1,3), label="var") + ggtitle("TGATAA excluded")
```

```{r linear modelling}

ngatan.by.peak[sort(rep(1:nrow(ngatan.by.peak),3)),] -> ngatan.repeated
ap10185 = ap[mappings]
bigdat = data.frame(mcols(subset(
  ap10185,
  select = c(peak, LE_nonNormed, L1_nonNormed, L3_nonNormed)
)))
melt(
  bigdat,
  id.vars = 'peak',
  variable.name = 'stage',
  value.name = 'signal'
) -> melted

melted[melted$stage == 'LE_nonNormed','stageVal'] <- 0
melted[melted$stage == 'L1_nonNormed','stageVal'] <- 1
melted[melted$stage == 'L3_nonNormed','stageVal'] <- 3

melted[melted$stage == 'LE_nonNormed','stageVal2'] <- 0
melted[melted$stage == 'L1_nonNormed','stageVal2'] <- 1
melted[melted$stage == 'L3_nonNormed','stageVal2'] <- 2


melted$TGATAA = c(ngatan.by.peak[, 'TGATAA'], ngatan.by.peak[, 'TGATAA'], ngatan.by.peak[, 'TGATAA'])
melted$TGATAT = c(ngatan.by.peak[, 'TGATAT'], ngatan.by.peak[, 'TGATAT'], ngatan.by.peak[, 'TGATAT'])
melted$AGATAA = c(ngatan.by.peak[, 'AGATAA'], ngatan.by.peak[, 'AGATAA'], ngatan.by.peak[, 'AGATAA'])

mod1 = lm(signal ~ stageVal, data=melted)
summary(mod1)

mod2 = lm(signal ~ stageVal*TGATAA, data=melted)
summary(mod2)

mod3 = lm(signal ~ stageVal*TGATAA + stageVal*AGATAA,data=melted)
summary(mod3)
```



``` {r co-occurance}

library(ymse)

ngatan.cooc = pairwise(ngatan.by.peak, function(x,y) sum(x*y > 0))

```