---
title: "RWC23_ELT2_ChIP_Boeck_Time_Resolved_RNA"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
library(biomaRt)
library(tidyverse)
library(readxl)
```



# Import Time-resolved RNA

```{r}
time_resolved_rna <- read.delim("./01_input/9_Boeck_et_al_2016_time-resolved_transcriptome/Unified_dcpm_per_wormbase_gene.txt", quote="", stringsAsFactors=FALSE)

paramart <- useMart("parasite_mart", dataset = "wbps_gene", host = "https://parasite.wormbase.org", port = 443)

time_resolved_rna <- getBM(
  mart = paramart,
  filter = c("wormbase_gseqname"),
  value = time_resolved_rna$WormbaseName,
  attributes = c("wormbase_gseq","wbps_gene_id", "wikigene_name")
) %>% right_join(time_resolved_rna, by = c("wormbase_gseq" = "WormbaseName"))


time_resolved_rna <- time_resolved_rna %>% drop_na(wbps_gene_id)
```


# Load in public intestine expression datasets

Initalize an empty dataframe.

```{r}
intestine_rna <- data.frame(WBGeneID = time_resolved_rna$wbps_gene_id, Sequence.name = time_resolved_rna$wormbase_gseq, Public_name = time_resolved_rna$wikigene_name)
```



# A Transcriptional Lineage of the Early C . elegans Embryo.
## Tintori SC, Nishimura EO, Golden P, Lieb JD, Goldstein B. Dev Cell. 2016;38(4):430-444. doi:10.1101/047746

```{r}
# Transcripts from embryonic 2Ep intestine from Sophie's single cell RNA-seq paper. See `intestine_tfs.Rmd`
tintori_2Ep <- read.csv("./01_input/7_Tintori_et_al_2016_Embryo_scRNA-seq/16cell_vs_Ep_p1_00.csv", header = FALSE, stringsAsFactors = FALSE)
tintori_columns <- c("gene", "p.value", "log.CPM", "log.FC")
colnames(tintori_2Ep) <- str_c("tintori_2Ep_", tintori_columns)
tintori_2Ep <- tintori_2Ep %>% mutate(tintori_2Ep_log.FC = tintori_2Ep_log.FC*-1)
intestine_rna <- intestine_rna %>% left_join(tintori_2Ep, by = c("Sequence.name" = "tintori_2Ep_gene"))

```

# Spatiotemporal transcriptomics reveals the evolutionary history of the endoderm germ layer.
## Hashimshony T, Feder M, Levin M, Hall BK, Yanai I.  Nature. 2015;519(7542):219-222. doi:10.1038/nature13996

```{r}
hashimshony_embryo <- read.csv("./01_input/1_Hashimshony_et_al_2015_CEL-seq/Hashimshony_Endoderm_Genes.csv", header = TRUE)
colnames(hashimshony_embryo) <- c("hashimshony_embryo_genes")

```

# Chromosomal clustering and GATA transcriptional regulation of intestine-expressed genes in C. elegans.
## Pauli F, Liu Y, Kim Y a, Chen P-J, Kim SK. Development. 2006;133(2):287-295. doi:10.1242/dev.02185

```{r}
# TFs from Pauli et al 
pauliGenes <- read_excel("./01_input/4_Pauli_et_al_2006_PAB-1_Pulldown_microarray/SupTable1_intestine_enriched_genes.xlsx")
colnames(pauliGenes) <- str_c("pauli_L4_", c("gene", "percentile", "pvalue", "description"))
# add to master datatable
intestine_rna <- intestine_rna %>% left_join(pauliGenes, by = c("Sequence.name" = "pauli_L4_gene"))
```

# A spatial and temporal map of C. elegans gene expression. Genome Research, 21(2), 325–41. https://doi.org/10.1101/gr.114595.110 
## Spencer, W. C., Zeller, G., Watson, J. D., Henz, S. R., Watkins, K. L., McWhirter, R. D., … Miller, D. M. (2011).

```{r}

# Late embryo intestine from Spencer et al
spencerLEgenes <- read.table("./01_input/6_Spencer_et_al_2010_FACS_and_pulldown_tilling_array/LE-intestine_enr_vs_ref.WS200.txt", quote="\"", comment.char="", header = TRUE)
colnames(spencerLEgenes) <- str_c("spencer_LE_", colnames(spencerLEgenes))
spencer_LE_subset <- spencerLEgenes %>% select(spencer_LE_ID, spencer_LE_AveExpr, spencer_LE_adj_P_Val, spencer_LE_FC)

# add to dataframe
intestine_rna <- merge(x = intestine_rna, y = spencer_LE_subset, by.x = 'WBGeneID', by.y = 'spencer_LE_ID', all.x = TRUE)



# L2 intestine genes from Spencer et al
spencerL2genes <- read.table("./01_input/6_Spencer_et_al_2010_FACS_and_pulldown_tilling_array/L2-intestine_enr_vs_ref.WS200.txt", quote="\"", comment.char="", header = TRUE)
colnames(spencerL2genes) <- str_c("spencer_L2_", colnames(spencerL2genes))
spencer_L2_subset <- spencerL2genes %>% select(spencer_L2_ID, spencer_L2_AveExpr, spencer_L2_adj_P_Val, spencer_L2_FC)

intestine_rna <- merge(x = intestine_rna, y = spencer_L2_subset, by.x = 'WBGeneID', by.y = 'spencer_L2_ID', all.x = TRUE)

```



# Filtering 1

```{r}
# # Tintori
# intestine_rna <-
#   intestine_rna %>% mutate(
#     tintori_2Ep_bool = if_else(
#       intestine_rna$WBGeneID %in%
#         (tintori_2Ep_enriched %>% filter(tintori_2Ep_log.FC > 1 & tintori_2Ep_p.value <= 0.05))$WBGeneID,
#       TRUE,
#       FALSE
#     )
#   )
# nrow(intestine_rna %>% filter(tintori_2Ep_bool == TRUE))
# # Hashimshony
# intestine_rna <- intestine_rna %>% mutate(hashimshony_embryo_bool = if_else(intestine_rna$Public_name %in% hashimshony_embryo$hashimshony_embryo_genes, TRUE, FALSE))
# nrow(intestine_rna %>% filter(hashimshony_embryo_bool == TRUE))
# 
# # Pauli
# # intestine_rna <- intestine_rna %>% mutate(pauli_L4_bool = if_else(intestine_rna$Sequence.name %in% (pauliGenes %>% filter(pauli_L4_percentile > 0.8))$pauli_L4_gene, TRUE, FALSE))
# intestine_rna <-
#   intestine_rna %>% mutate(pauli_L4_bool = if_else(
#     intestine_rna$Sequence.name %in% pauliGenes$pauli_L4_gene,
#     TRUE,
#     FALSE
#   ))
# nrow(intestine_rna %>% filter(pauli_L4_bool == TRUE))
# 
# # Spencer Late Embryo
# intestine_rna <- intestine_rna %>% mutate(spencer_LE_bool = if_else(intestine_rna$WBGeneID %in% spencer_LE_subset$spencer_LE_ID, TRUE, FALSE))
# nrow(intestine_rna %>% filter(spencer_LE_bool == TRUE))
# 
# # Spencer L2
# intestine_rna <- intestine_rna %>% mutate(spencer_L2_bool = if_else(intestine_rna$WBGeneID %in% spencer_L2_subset$spencer_L2_ID, TRUE, FALSE))
# nrow(intestine_rna %>% filter(spencer_L2_bool == TRUE))
```

# Filtering 2

```{r}
# # Tintori
# intestine_rna <-
#   intestine_rna %>% mutate(
#     tintori_2Ep_bool = if_else(
#       intestine_rna$WBGeneID %in% 
#         (tintori_2Ep_enriched %>% filter(tintori_2Ep_log.FC > 5 & tintori_2Ep_p.value <= 0.05))$WBGeneID,
#       TRUE,
#       FALSE
#     )
#   )
# nrow(intestine_rna %>% filter(tintori_2Ep_bool == TRUE))
# 
# # Hashimshony
# intestine_rna <- intestine_rna %>% mutate(hashimshony_embryo_bool = if_else(intestine_rna$Public_name %in% hashimshony_embryo$hashimshony_embryo_genes, TRUE, FALSE))
# nrow(intestine_rna %>% filter(hashimshony_embryo_bool == TRUE))
# 
# # Pauli
# intestine_rna <-
#   intestine_rna %>% mutate(pauli_L4_bool = if_else(
#     intestine_rna$Sequence.name %in% (pauliGenes %>% filter(pauli_L4_percentile > 0.8))$pauli_L4_gene,
#     TRUE,
#     FALSE
#   ))
# nrow(intestine_rna %>% filter(pauli_L4_bool == TRUE))
# 
# # Spencer Late Embryo
# intestine_rna <-
#   intestine_rna %>% mutate(
#     spencer_LE_bool = if_else(
#       intestine_rna$WBGeneID %in% (spencer_LE_subset %>% filter(spencer_LE_FC >= 5))$spencer_LE_ID,
#       TRUE,
#       FALSE
#     )
#   )
# nrow(intestine_rna %>% filter(spencer_LE_bool == TRUE))
# 
# # Spencer L2
# intestine_rna <-
#   intestine_rna %>% mutate(
#     spencer_L2_bool = if_else(
#       intestine_rna$WBGeneID %in% (spencer_L2_subset %>% filter(spencer_L2_FC >= 5))$spencer_L2_ID,
#       TRUE,
#       FALSE
#     )
#   )
# nrow(intestine_rna %>% filter(spencer_L2_bool == TRUE))
```

# Filtering 3

```{r}
# # Tintori
# intestine_rna <-
#   intestine_rna %>% mutate(
#     tintori_2Ep_bool = if_else(
#       intestine_rna$WBGeneID %in% 
#         (tintori_2Ep_enriched %>% filter(tintori_2Ep_log.FC > 8 & tintori_2Ep_p.value <= 0.05))$WBGeneID,
#       TRUE,
#       FALSE
#     )
#   )
# nrow(intestine_rna %>% filter(tintori_2Ep_bool == TRUE))
# 
# # Hashimshony
# intestine_rna <- intestine_rna %>% mutate(hashimshony_embryo_bool = if_else(intestine_rna$Public_name %in% hashimshony_embryo$hashimshony_embryo_genes, TRUE, FALSE))
# nrow(intestine_rna %>% filter(hashimshony_embryo_bool == TRUE))
# 
# # Pauli
# intestine_rna <-
#   intestine_rna %>% mutate(pauli_L4_bool = if_else(
#     intestine_rna$Sequence.name %in% (pauliGenes %>% filter(pauli_L4_percentile > 0.9))$pauli_L4_gene,
#     TRUE,
#     FALSE
#   ))
# nrow(intestine_rna %>% filter(pauli_L4_bool == TRUE))
# 
# # Spencer Late Embryo
# intestine_rna <-
#   intestine_rna %>% mutate(
#     spencer_LE_bool = if_else(
#       intestine_rna$WBGeneID %in% (spencer_LE_subset %>% filter(spencer_LE_FC >= 8))$spencer_LE_ID,
#       TRUE,
#       FALSE
#     )
#   )
# nrow(intestine_rna %>% filter(spencer_LE_bool == TRUE))
# 
# # Spencer L2
# intestine_rna <-
#   intestine_rna %>% mutate(
#     spencer_L2_bool = if_else(
#       intestine_rna$WBGeneID %in% (spencer_L2_subset %>% filter(spencer_L2_FC >= 6))$spencer_L2_ID,
#       TRUE,
#       FALSE
#     )
#   )
# nrow(intestine_rna %>% filter(spencer_L2_bool == TRUE))
```
# Final Filtering

```{r}
# Tintori
intestine_rna <-
  intestine_rna %>% mutate(
    tintori_2Ep_bool = if_else(
      intestine_rna$WBGeneID %in%
        (tintori_2Ep_enriched %>% filter(tintori_2Ep_log.FC > 5 & tintori_2Ep_p.value <= 0.05))$WBGeneID,
      TRUE,
      FALSE
    )
  )
nrow(intestine_rna %>% filter(tintori_2Ep_bool == TRUE))
# Hashimshony
intestine_rna <- intestine_rna %>% mutate(hashimshony_embryo_bool = if_else(intestine_rna$Public_name %in% hashimshony_embryo$hashimshony_embryo_genes, TRUE, FALSE))
nrow(intestine_rna %>% filter(hashimshony_embryo_bool == TRUE))

# Pauli
# intestine_rna <- intestine_rna %>% mutate(pauli_L4_bool = if_else(intestine_rna$Sequence.name %in% (pauliGenes %>% filter(pauli_L4_percentile > 0.8))$pauli_L4_gene, TRUE, FALSE))
intestine_rna <-
  intestine_rna %>% mutate(pauli_L4_bool = if_else(
    intestine_rna$Sequence.name %in% pauliGenes$pauli_L4_gene,
    TRUE,
    FALSE
  ))
nrow(intestine_rna %>% filter(pauli_L4_bool == TRUE))

# Spencer Late Embryo
intestine_rna <- intestine_rna %>% mutate(spencer_LE_bool = if_else(intestine_rna$WBGeneID %in% spencer_LE_subset$spencer_LE_ID, TRUE, FALSE))
nrow(intestine_rna %>% filter(spencer_LE_bool == TRUE))

# Spencer L2
intestine_rna <- intestine_rna %>% mutate(spencer_L2_bool = if_else(intestine_rna$WBGeneID %in% spencer_L2_subset$spencer_L2_ID, TRUE, FALSE))
nrow(intestine_rna %>% filter(spencer_L2_bool == TRUE))
```
```{r}
intestine_rna %>% filter(Public_name %in% c("M88.4", "bcl-11", "sre-23", "T25D10.4", "aat-6", "pgp-1", "ftn-1", "elt-2", "elt-7")) %>% dplyr::select(Public_name, contains("bool"))
```

```{r}
# write.csv(
#   file = "./02_output/smFISH_gene_Values.csv",
#   intestine_rna %>% filter(
#     Public_name %in% c(
#       "M88.4",
#       "bcl-11",
#       "sre-23",
#       "T25D10.4",
#       "aat-6",
#       "pgp-1",
#       "ftn-1",
#       "elt-2",
#       "elt-7"
#     )
#   ) %>% dplyr::select(
#     Public_name:spencer_L2_FC,-contains("bool"),-contains("description")
#   )
# )
```


Make a vector that contains intestine expressed genes from any stage

```{r}
intestine_gene_list <- intestine_rna %>% filter(
  tintori_2Ep_bool == TRUE |
    hashimshony_embryo_bool == TRUE |
    pauli_L4_bool == TRUE |
    spencer_LE_bool == TRUE |
    spencer_L2_bool == TRUE
) %>% dplyr::select(WBGeneID, Public_name)

dim(intestine_gene_list)
```

Write output files

```{r}
write_csv(intestine_gene_list, "./02_output/RWC23_Public_Intestine_RNA_Gene_List.csv")
write_csv(intestine_rna, "./02_output/RWC23_Public_Intestine_RNA_Full_Data.csv")
```


# UpSet plot of overlaps

```{r}
# there are 5 sets of interest
# set up the combination matrix
intestine_upset <-
  intestine_gene_list %>% left_join(intestine_rna, by = "WBGeneID") %>% dplyr::select(WBGeneID, contains("bool")) %>% column_to_rownames(var = "WBGeneID") %>% dplyr::rename(
    tint_2Ep = "tintori_2Ep_bool",
    hash_emb = "hashimshony_embryo_bool",
    pauli_L4 = "pauli_L4_bool",
    spen_LE = "spencer_LE_bool",
    spen_L2 = "spencer_L2_bool"
  )
head(intestine_upset)
nrow(intestine_upset)
# replace TRUE/FALSE with 1/0
intestine_upset[, sapply(intestine_upset, is.logical)] <-
  lapply(intestine_upset[, sapply(intestine_upset, is.logical)], as.numeric)
# convert to matrix
intestine_upset_matrix <- as.matrix.data.frame(intestine_upset)
# peek
head(intestine_upset_matrix)
# number of rows
nrow(intestine_upset_matrix)
```

```{r}
# install.packages("UpSetR")
library(UpSetR)
library(ComplexHeatmap)

m1 <- make_comb_mat(intestine_upset_matrix)
m1
```
```{r}
UpSet(m1, 
      set_order = c("tint_2Ep", "hash_emb", "spen_LE", "spen_L2", "pauli_L4"),
      comb_order = order(comb_size(m1)))
```


# Metadata

Initalize a seprate empty "metadata" dataframe to store descriptive information

```{r}
RWC23_Metadata <- data.frame(column_ID = "", column_detail = "", source_DOI = "", source_detail = "", seq_method = "")

# Tintori metadata

RWC23_Metadata <- RWC23_Metadata %>% add_row(
  column_ID = colnames(tintori_2Ep[1]),
  column_detail = "Gene sequence name",
  source_DOI = "10.1016/j.devcel.2016.07.025",
  source_detail = "Single cell RNA-seq of hand dissected embryo. Accessed from tintori.bio.unc.edu.",
  seq_method = "Illumina"
  ) %>% add_row(
  column_ID = colnames(tintori_2Ep[2]),
  column_detail = "Gene p-value",
  source_DOI = "10.1016/j.devcel.2016.07.025",
  source_detail = "Single cell RNA-seq of hand dissected embryo. Accessed from tintori.bio.unc.edu.",
  seq_method = "Illumina"
  ) %>% add_row(
  column_ID = colnames(tintori_2Ep[3]),
  column_detail = "Counts per transcript per million mapped reads.",
  source_DOI = "10.1016/j.devcel.2016.07.025",
  source_detail = "Single cell RNA-seq of hand dissected embryo. Accessed from tintori.bio.unc.edu.",
  seq_method = "Illumina"
  ) %>% add_row(
  column_ID = colnames(tintori_2Ep[4]),
  column_detail = "Log2(Fold Change) of gene in the 2Ep cell when compared to the whole embryo. Less than zero is enriched in the 1E cell. Greater than zero is enriched in the whole 16 cell embryo",
  source_DOI = "10.1016/j.devcel.2016.07.025",
  source_detail = "Single cell RNA-seq of hand dissected embryo. Accessed from tintori.bio.unc.edu.",
  seq_method = "Illumina"
  )


# Hashimshony metadata

RWC23_Metadata <- RWC23_Metadata %>% add_row(
  column_ID = colnames(hashimshony_embryo[1]),
  column_detail = "Gene public name",
  source_DOI = "10.1038/nature13996",
  source_detail = "Supplementary Table 4. Endoderm gene set from Single cell RNA-seq of cultured dissociated embryos.",
  seq_method = "Illumina"
  )

# update metadata
RWC23_Metadata <- RWC23_Metadata  %>% add_row(
  column_ID = colnames(pauliGenes[1]),
  column_detail = "Sequence name",
  source_DOI = "10.1242/dev.02185",
  source_detail = "Supplementary Table 1. Microarray of ges-1P::PAB-1 pulldown in L4 worms"
  ) %>% add_row(
  column_ID = colnames(pauliGenes[2]),
  column_detail = "Percentile ranking of transcript abundance",
  source_DOI = "10.1242/dev.02185",
  source_detail = "Supplementary Table 1. Microarray of ges-1P::PAB-1 pulldown in L4 worms"
  ) %>% add_row(
  column_ID = colnames(pauliGenes[3]),
  column_detail = "Student's T-test to calculate significant enrichment of intestine transcripts",
  source_DOI = "10.1242/dev.02185",
  source_detail = "Supplementary Table 1. Microarray of ges-1P::PAB-1 pulldown in L4 worms"
  ) %>% add_row(
  column_ID = colnames(pauliGenes[4]),
  column_detail = "Short descriptions generated by WormBase",
  source_DOI = "10.1242/dev.02185",
  source_detail = "Supplementary Table 1. Microarray of ges-1P::PAB-1 pulldown in L4 worms"
  ) %>% add_row(
  column_ID = "pauli_L4_bool",
  column_detail = "True if included on the list, false if absent on the list",
  source_DOI = "10.1242/dev.02185",
  source_detail = "Supplementary Table 1. Microarray of ges-1P::PAB-1 pulldown in L4 worms"
  ) 

# add sequencing method column

RWC23_Metadata <- mutate(RWC23_Metadata, seq_method = ifelse(RWC23_Metadata$column_ID %in% c(colnames(pauliGenes), "pauli_L4_bool"), "Microarray",RWC23_Metadata$seq_method))


# Add Spencer metadata

RWC23_Metadata <- RWC23_Metadata  %>% add_row(
  column_ID = colnames(spencer_LE_subset[1]),
  column_detail = "Wormbase Gene ID",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  ) %>% add_row(
  column_ID = colnames(spencer_LE_subset[2]),
  column_detail = "??",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  ) %>% add_row(
  column_ID = colnames(spencer_LE_subset[3]),
  column_detail = "??",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  ) %>% add_row(
  column_ID = colnames(spencer_LE_subset[4]),
  column_detail = "??",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  ) %>% add_row(
  column_ID = "spencer_LE_bool",
  column_detail = "True if present in this list",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  )

RWC23_Metadata <- mutate(RWC23_Metadata, seq_method = ifelse(RWC23_Metadata$column_ID %in% c(colnames(spencer_LE_subset), "spencer_LE_bool"), "Microarray",RWC23_Metadata$seq_method))

RWC23_Metadata <- RWC23_Metadata  %>% add_row(
  column_ID = colnames(spencer_L2_subset[1]),
  column_detail = "Wormbase Gene ID",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  ) %>% add_row(
  column_ID = colnames(spencer_L2_subset[2]),
  column_detail = "??",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  ) %>% add_row(
  column_ID = colnames(spencer_L2_subset[3]),
  column_detail = "??",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  ) %>% add_row(
  column_ID = colnames(spencer_L2_subset[4]),
  column_detail = "??",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  ) %>% add_row(
  column_ID = "spencer_L2_bool",
  column_detail = "True if present in this list",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  )

RWC23_Metadata <- mutate(RWC23_Metadata, seq_method = ifelse(RWC23_Metadata$column_ID %in% c(colnames(spencer_L2_subset), "spencer_L2_bool"), "Microarray",RWC23_Metadata$seq_method))
```

# write output metadata file

```{r}
library(openxlsx)
# write.xlsx(RWC23_Metadata, "./02_output/RWC23_Public_Intestine_RNA_Metadata.xlsx")

```

