---
title: "KW_ELT2_Figure3"
author: "Robert Williams"
date: "3/3/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Install packages
```{r}
# BiocManager::install("biomaRt")
# install.packages("tidyverse")
# install.packages("readxl")
# install.packages("openxlsx")
# BiocManager::install("ComplexHeatmap")
# BiocManager::install("GenomicRanges")
```

Load packages
```{r}
library(biomaRt)
library(tidyverse)
library(readxl)
library(ComplexHeatmap)
library(lubridate)
```

Load custom functions

```{r}
source("../RWC23_Functions.R")
```

Set plot variable

```{r}
plot <- FALSE
plotdir <- "./03_plots/"
```

# TODO

- 210616
  - add blaize to metadata
  - soft/med/hard variance cutoff
- 210622
  - Add Early, middle, late embryo intestine
  - Calculate unified tpm from counts
  - Time resolved heatmap with different threshold cutoffs (make plotting function)

# Import Time-resolved RNA

```{r}
time_resolved_rna <- read.delim("./01_input/9_Boeck_et_al_2016_time-resolved_transcriptome/Unified_dcpm_per_wormbase_gene.txt", quote="", stringsAsFactors=FALSE)

paramart <- useMart("parasite_mart", dataset = "wbps_gene", host = "https://parasite.wormbase.org", port = 443)

time_resolved_rna <- getBM(
  mart = paramart,
  filter = c("wormbase_gseqname"),
  value = time_resolved_rna$WormbaseName,
  attributes = c("wormbase_gseq","wbps_gene_id", "wikigene_name")
) %>% right_join(time_resolved_rna, by = c("wormbase_gseq" = "WormbaseName"))


time_resolved_rna <- time_resolved_rna %>% drop_na(wbps_gene_id)
```


# Load in public intestine expression datasets

Initalize an empty dataframe.

```{r}
intestine_rna <- data.frame(WBGeneID = time_resolved_rna$wbps_gene_id, Sequence.name = time_resolved_rna$wormbase_gseq, Public_name = time_resolved_rna$wikigene_name)
```


# Comparative RNA-Seq Analysis Reveals Pervasive Tissue-Specific Alternative Polyadenylation in Caenorhabditis Elegans Intestine and Muscles.
## Blazie, Stephen M, Cody Babb, Henry Wilky, Alan Rawls, Jin G Park, and Marco Mangone. https://doi.org/10.1186/s12915-015-0116-6


```{r}
blazie_mixed <- read_xlsx(path = "./01_input/2_Blazie_et_al_2015_PAB-1_Pulldown_RNA-seq/Blazie_et_al_2015_Table_S3.xlsx", sheet = "ges-1 - Unique Genes", col_names = TRUE) %>% dplyr::rename(WBGeneID = `WB Gene ID`) %>% dplyr::select(-Description)

colnames(blazie_mixed) <- str_c("blazie_", colnames(blazie_mixed))

head(blazie_mixed)
hist(blazie_mixed$blazie_FPKM)
intestine_rna<- intestine_rna %>% left_join(blazie_mixed, by = c("Sequence.name" = "blazie_Cosmid"))
head(intestine_rna)
```

# Spatiotemporal transcriptomics reveals the evolutionary history of the endoderm germ layer.
## Hashimshony T, Feder M, Levin M, Hall BK, Yanai I.  Nature. 2015;519(7542):219-222. doi:10.1038/nature13996

```{r}
hashimshony_embryo <- read.csv("./01_input/1_Hashimshony_et_al_2015_CEL-seq/Hashimshony_Endoderm_Genes.csv", header = TRUE)
colnames(hashimshony_embryo) <- c("hashimshony_embryo_genes")

```

# Chromosomal clustering and GATA transcriptional regulation of intestine-expressed genes in C. elegans.
## Pauli F, Liu Y, Kim Y a, Chen P-J, Kim SK. Development. 2006;133(2):287-295. doi:10.1242/dev.02185

```{r}
# TFs from Pauli et al 
pauliGenes <- read_excel("./01_input/4_Pauli_et_al_2006_PAB-1_Pulldown_microarray/SupTable1_intestine_enriched_genes.xlsx")
colnames(pauliGenes) <- str_c("pauli_L4_", c("gene", "percentile", "pvalue", "description"))
# add to master datatable
intestine_rna <- intestine_rna %>% left_join(pauliGenes, by = c("Sequence.name" = "pauli_L4_gene"))
```

# A spatial and temporal map of C. elegans gene expression. Genome Research, 21(2), 325–41. https://doi.org/10.1101/gr.114595.110 
## Spencer, W. C., Zeller, G., Watson, J. D., Henz, S. R., Watkins, K. L., McWhirter, R. D., … Miller, D. M. (2011).

```{r}

# Late embryo intestine from Spencer et al
spencerLEgenes <- read.table("./01_input/6_Spencer_et_al_2010_FACS_and_pulldown_tilling_array/LE-intestine_enr_vs_ref.WS200.txt", quote="\"", comment.char="", header = TRUE)
colnames(spencerLEgenes) <- str_c("spencer_LE_", colnames(spencerLEgenes))
spencer_LE_subset <- spencerLEgenes %>% dplyr::select(spencer_LE_ID, spencer_LE_AveExpr, spencer_LE_adj_P_Val, spencer_LE_FC)

# add to dataframe
intestine_rna <- merge(x = intestine_rna, y = spencer_LE_subset, by.x = 'WBGeneID', by.y = 'spencer_LE_ID', all.x = TRUE)



# L2 intestine genes from Spencer et al
spencerL2genes <- read.table("./01_input/6_Spencer_et_al_2010_FACS_and_pulldown_tilling_array/L2-intestine_enr_vs_ref.WS200.txt", quote="\"", comment.char="", header = TRUE)
colnames(spencerL2genes) <- str_c("spencer_L2_", colnames(spencerL2genes))
spencer_L2_subset <- spencerL2genes %>% dplyr::select(spencer_L2_ID, spencer_L2_AveExpr, spencer_L2_adj_P_Val, spencer_L2_FC)

intestine_rna <- merge(x = intestine_rna, y = spencer_L2_subset, by.x = 'WBGeneID', by.y = 'spencer_L2_ID', all.x = TRUE)

```

# Intestine Gene Filtering

```{r}
# Blazie

intestine_rna <- intestine_rna %>% mutate(blazie_bool = if_else(intestine_rna$Sequence.name %in% blazie_mixed$blazie_Cosmid, TRUE, FALSE))
nrow(intestine_rna %>% filter(blazie_bool == TRUE))

# Hashimshony
intestine_rna <- intestine_rna %>% mutate(hashimshony_embryo_bool = if_else(intestine_rna$Public_name %in% hashimshony_embryo$hashimshony_embryo_genes, TRUE, FALSE))
nrow(intestine_rna %>% filter(hashimshony_embryo_bool == TRUE))

# Pauli
intestine_rna <-
  intestine_rna %>% mutate(pauli_L4_bool = if_else(
    intestine_rna$Sequence.name %in% pauliGenes$pauli_L4_gene,
    TRUE,
    FALSE
  ))
nrow(intestine_rna %>% filter(pauli_L4_bool == TRUE))

# Spencer Late Embryo
intestine_rna <- intestine_rna %>% mutate(spencer_LE_bool = if_else(intestine_rna$WBGeneID %in% spencer_LE_subset$spencer_LE_ID, TRUE, FALSE))
nrow(intestine_rna %>% filter(spencer_LE_bool == TRUE))

# Spencer L2
intestine_rna <- intestine_rna %>% mutate(spencer_L2_bool = if_else(intestine_rna$WBGeneID %in% spencer_L2_subset$spencer_L2_ID, TRUE, FALSE))
nrow(intestine_rna %>% filter(spencer_L2_bool == TRUE))
```
```{r}
intestine_rna %>% filter(Public_name %in% c("M88.4", "bcl-11", "sre-23", "T25D10.4", "aat-6", "pgp-1", "ftn-1", "elt-2", "elt-7")) %>% dplyr::select(Public_name, contains("bool"))
```


Make a vector that contains intestine expressed genes from any stage

```{r}
intestine_gene_list <- intestine_rna %>% filter(
    blazie_bool == TRUE |
    hashimshony_embryo_bool == TRUE |
    pauli_L4_bool == TRUE |
    spencer_LE_bool == TRUE |
    spencer_L2_bool == TRUE
) %>% dplyr::select(WBGeneID, Sequence.name, Public_name, contains("bool")) %>% dplyr::rename(
  blazie_mixed = "blazie_bool",
  hashimshony_emb = "hashimshony_embryo_bool",
  pauli_L4 = "pauli_L4_bool",
  spencer_LE = "spencer_LE_bool",
  spencer_L2 = "spencer_L2_bool"
)

head(intestine_gene_list)
dim(intestine_gene_list)
```


## Boeck expression values for genes of interest

```{r fig.width=4, fig.height=7}
smFISH_genes <- c("M88.4", "bcl-11", "sre-23", "T25D10.4", "aat-6", "pgp-1", "ftn-1", "elt-2", "elt-7")

time_resolved_rna %>% filter(wikigene_name %in% smFISH_genes) %>% pivot_longer(cols = emb_316min:YA, names_to = "Stage",values_to = "Normalized_Expression") %>%
  ggplot(aes(x = Stage, y = Normalized_Expression)) +
  geom_point() +
  facet_grid(wikigene_name~.) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
  ggtitle("Time-Resolved RNA Expression of\nIntestine Genes for smFISH") +
  ylab("depth of coverage per base per million reads (dcpm)")
```
```{r fig.width=3.25, fig.height=5}

elt2elt7_dcpm <- time_resolved_rna %>% filter(wikigene_name %in% c("elt-2", "elt-7")) %>% pivot_longer(cols = emb_316min:YA, names_to = "Stage",values_to = "Normalized_Expression") %>%
  ggplot(aes(x = Stage, y = Normalized_Expression)) +
  geom_point() +
  facet_grid(wikigene_name~.) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
  ggtitle("Time-Resolved RNA Expression") +
  ylab("depth of coverage per base per million reads (dcpm)")
elt2elt7_dcpm
# ggsave(elt2elt7_dcpm, file = "03_plots/210316_elt2elt7_timeresolvedRNA.pdf", width = 3.25, height = 5)
```

## Write output files

```{r}
# write_csv(intestine_gene_list, "./02_output/Fig3_Public_Intestine_RNA_Gene_List.csv")
# write_csv(intestine_rna, "./02_output/Fig3_Public_Intestine_RNA_Full_Data.csv")
```

Read in files for downstream work

```{r}
intestine_rna <- read.csv("./02_output/Fig3_Public_Intestine_RNA_Full_Data.csv", stringsAsFactors = FALSE)
intestine_gene_list <- read.csv("./02_output/Fig3_Public_Intestine_RNA_Gene_List.csv", stringsAsFactors = FALSE)
```


# UpSet plot of overlaps

```{r}
# there are 5 sets of interest
# set up the combination matrix
intestine_upset <-
  intestine_gene_list %>% left_join(intestine_rna, by = "WBGeneID") %>% dplyr::select(WBGeneID, contains("bool")) %>% column_to_rownames(var = "WBGeneID") %>% dplyr::rename(
    blazie_mixed = "blazie_bool",
    hash_emb = "hashimshony_embryo_bool",
    pauli_L4 = "pauli_L4_bool",
    spen_LE = "spencer_LE_bool",
    spen_L2 = "spencer_L2_bool"
  )
head(intestine_upset)
nrow(intestine_upset)
# replace TRUE/FALSE with 1/0
intestine_upset[, sapply(intestine_upset, is.logical)] <-
  lapply(intestine_upset[, sapply(intestine_upset, is.logical)], as.numeric)
# convert to matrix
intestine_upset_matrix <- as.matrix.data.frame(intestine_upset)
# peek
head(intestine_upset_matrix)
# number of rows
nrow(intestine_upset_matrix)
```

```{r}
m1 <- make_comb_mat(intestine_upset_matrix)
m1
```
## The total number of intestine genes is `r sum(comb_size(m1))`

```{r}
sum(comb_size(m1))
```


```{r fig.width=5, fig.height=3}
intestine_upset_plot <- UpSet(m1, 
      set_order = c("blazie_mixed","hash_emb", "spen_LE", "spen_L2", "pauli_L4"),
      comb_order = order(comb_size(m1)))
intestine_upset_plot

if(plot == TRUE){
  myPDFplot(intestine_upset_plot, name = "Intestine_Gene_Upset_Plot", height = 3, width = 5,plotdir)
}
```


# Metadata

Initalize a seprate empty "metadata" dataframe to store descriptive information

# TODO: add blaize to metadata

```{r}
RWC23_Metadata <- data.frame(column_ID = "", column_detail = "", source_DOI = "", source_detail = "", seq_method = "")

# Tintori metadata

# RWC23_Metadata <- RWC23_Metadata %>% add_row(
#   column_ID = colnames(tintori_2Ep[1]),
#   column_detail = "Gene sequence name",
#   source_DOI = "10.1016/j.devcel.2016.07.025",
#   source_detail = "Single cell RNA-seq of hand dissected embryo. Accessed from tintori.bio.unc.edu.",
#   seq_method = "Illumina"
#   ) %>% add_row(
#   column_ID = colnames(tintori_2Ep[2]),
#   column_detail = "Gene p-value",
#   source_DOI = "10.1016/j.devcel.2016.07.025",
#   source_detail = "Single cell RNA-seq of hand dissected embryo. Accessed from tintori.bio.unc.edu.",
#   seq_method = "Illumina"
#   ) %>% add_row(
#   column_ID = colnames(tintori_2Ep[3]),
#   column_detail = "Counts per transcript per million mapped reads.",
#   source_DOI = "10.1016/j.devcel.2016.07.025",
#   source_detail = "Single cell RNA-seq of hand dissected embryo. Accessed from tintori.bio.unc.edu.",
#   seq_method = "Illumina"
#   ) %>% add_row(
#   column_ID = colnames(tintori_2Ep[4]),
#   column_detail = "Log2(Fold Change) of gene in the 2Ep cell when compared to the whole embryo. Less than zero is enriched in the 1E cell. Greater than zero is enriched in the whole 16 cell embryo",
#   source_DOI = "10.1016/j.devcel.2016.07.025",
#   source_detail = "Single cell RNA-seq of hand dissected embryo. Accessed from tintori.bio.unc.edu.",
#   seq_method = "Illumina"
#   )


# Hashimshony metadata

RWC23_Metadata <- RWC23_Metadata %>% add_row(
  column_ID = colnames(hashimshony_embryo[1]),
  column_detail = "Gene public name",
  source_DOI = "10.1038/nature13996",
  source_detail = "Supplementary Table 4. Endoderm gene set from Single cell RNA-seq of cultured dissociated embryos.",
  seq_method = "Illumina"
  )

# update metadata
RWC23_Metadata <- RWC23_Metadata  %>% add_row(
  column_ID = colnames(pauliGenes[1]),
  column_detail = "Sequence name",
  source_DOI = "10.1242/dev.02185",
  source_detail = "Supplementary Table 1. Microarray of ges-1P::PAB-1 pulldown in L4 worms"
  ) %>% add_row(
  column_ID = colnames(pauliGenes[2]),
  column_detail = "Percentile ranking of transcript abundance",
  source_DOI = "10.1242/dev.02185",
  source_detail = "Supplementary Table 1. Microarray of ges-1P::PAB-1 pulldown in L4 worms"
  ) %>% add_row(
  column_ID = colnames(pauliGenes[3]),
  column_detail = "Student's T-test to calculate significant enrichment of intestine transcripts",
  source_DOI = "10.1242/dev.02185",
  source_detail = "Supplementary Table 1. Microarray of ges-1P::PAB-1 pulldown in L4 worms"
  ) %>% add_row(
  column_ID = colnames(pauliGenes[4]),
  column_detail = "Short descriptions generated by WormBase",
  source_DOI = "10.1242/dev.02185",
  source_detail = "Supplementary Table 1. Microarray of ges-1P::PAB-1 pulldown in L4 worms"
  ) %>% add_row(
  column_ID = "pauli_L4_bool",
  column_detail = "True if included on the list, false if absent on the list",
  source_DOI = "10.1242/dev.02185",
  source_detail = "Supplementary Table 1. Microarray of ges-1P::PAB-1 pulldown in L4 worms"
  ) 

# add sequencing method column

RWC23_Metadata <- mutate(RWC23_Metadata, seq_method = ifelse(RWC23_Metadata$column_ID %in% c(colnames(pauliGenes), "pauli_L4_bool"), "Microarray",RWC23_Metadata$seq_method))


# Add Spencer metadata

RWC23_Metadata <- RWC23_Metadata  %>% add_row(
  column_ID = colnames(spencer_LE_subset[1]),
  column_detail = "Wormbase Gene ID",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  ) %>% add_row(
  column_ID = colnames(spencer_LE_subset[2]),
  column_detail = "??",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  ) %>% add_row(
  column_ID = colnames(spencer_LE_subset[3]),
  column_detail = "??",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  ) %>% add_row(
  column_ID = colnames(spencer_LE_subset[4]),
  column_detail = "??",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  ) %>% add_row(
  column_ID = "spencer_LE_bool",
  column_detail = "True if present in this list",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  )

RWC23_Metadata <- mutate(RWC23_Metadata, seq_method = ifelse(RWC23_Metadata$column_ID %in% c(colnames(spencer_LE_subset), "spencer_LE_bool"), "Microarray",RWC23_Metadata$seq_method))

RWC23_Metadata <- RWC23_Metadata  %>% add_row(
  column_ID = colnames(spencer_L2_subset[1]),
  column_detail = "Wormbase Gene ID",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  ) %>% add_row(
  column_ID = colnames(spencer_L2_subset[2]),
  column_detail = "??",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  ) %>% add_row(
  column_ID = colnames(spencer_L2_subset[3]),
  column_detail = "??",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  ) %>% add_row(
  column_ID = colnames(spencer_L2_subset[4]),
  column_detail = "??",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  ) %>% add_row(
  column_ID = "spencer_L2_bool",
  column_detail = "True if present in this list",
  source_DOI = "10.1101/gr.114595.110",
  source_detail = "FACS Sorted ELT-2::GFP worms. https://www.vanderbilt.edu/wormdoc/wormmap"
  )

RWC23_Metadata <- mutate(RWC23_Metadata, seq_method = ifelse(RWC23_Metadata$column_ID %in% c(colnames(spencer_L2_subset), "spencer_L2_bool"), "Microarray",RWC23_Metadata$seq_method))
```

## write output metadata file

```{r}
# write.xlsx(RWC23_Metadata, "./02_output/Fig3_Public_Intestine_RNA_Metadata.xlsx")
```


# Import ELT-2 ChIP-seq binding data

Make sure annotatedPeaks file is linked to `01_input` directory.  
The path is: `../../../ELT2_binding/annotatedPeaks.rds`

```{r}
elt2_GRange <- readRDS("01_input/annotatedPeaks.rds")
head(elt2_GRange)
elt2_peaks <- as.data.frame(mcols(elt2_GRange))
elt2_peaks <- elt2_peaks %>% dplyr::rename(cluster.description = k4labels, WBGeneID = feature)
elt2_cluster_names <- c("Embryo_Specific",
                        "Larval",
                        "Increasing",
                        "L3_High",
                        "Not_Changing")

elt2_peaks$cluster.description <-
  factor(
    elt2_peaks$cluster.description,
    levels = c(
      "LE-specific",
      "Post-embryonic",
      "Increasing",
      "L3-high",
      "Not-changing or not IDR-passing"
    ),
    labels = elt2_cluster_names
  )

elt2_peaks %>% head
```
```{r}
table(elt2_peaks$cluster.description)
```

# Select genes with one ELT-2 occupancy pattern category

```{r}
elt2_peaks_wider <- elt2_peaks %>% group_by(WBGeneID, cluster.description) %>% summarise(cluster.count = n()) %>% pivot_wider(id_cols = WBGeneID, names_from = cluster.description, values_from = cluster.count, values_fill = 0) %>% ungroup()

```

```{r}
elt2_gene_occupancy <- elt2_peaks_wider %>%
  mutate(
    single_cluster =
      case_when(
        Increasing >= 1 &
          L3_High == 0 &
          Larval == 0 &
          Not_Changing == 0 &
          Embryo_Specific == 0 ~ "Increasing",
        Increasing == 0 &
          L3_High >= 1 &
          Larval == 0 &
          Not_Changing == 0 &
          Embryo_Specific == 0 ~ "L3_High",
        Increasing == 0 &
          L3_High == 0 &
          Larval >= 1 &
          Not_Changing == 0 &
          Embryo_Specific == 0 ~ "Larval",
        Increasing == 0 &
          L3_High == 0 &
          Larval == 0 &
          Not_Changing >= 1 &
          Embryo_Specific == 0 ~ "Not_Changing",
        Increasing == 0 &
          L3_High == 0 &
          Larval == 0 &
          Not_Changing == 0 &
          Embryo_Specific >= 1 ~ "Embryo_Specific",
        TRUE ~ "Multi_Cluster"
      )
  )
    
single_cluster_genes <- elt2_gene_occupancy %>%
  dplyr::select(WBGeneID, single_cluster) %>%
  filter(single_cluster != "Multi_Cluster")

head(single_cluster_genes)
nrow(single_cluster_genes)
table(elt2_gene_occupancy$single_cluster)
```

```{r}

boeck_count_mat <- time_resolved_rna %>% 
  column_to_rownames(var = "wbps_gene_id") %>%
  dplyr::select(emb_510min:emb_665min,L1:YA) %>% as.matrix.data.frame()

```
Plot the variance vs the mean

```{r}
var_mean_plot <- function(mat, meancut, varcut, title = NULL){
 mat_var <- apply(mat, 1, var)
 mat_mean <- apply(mat, 1, mean)
 plot(log10(mat_mean), log10(mat_var), pch='.', main = title)
abline(h=log10(varcut), col='red')
abline(v=log10(meancut), col='red')
}
var_mean_plot(boeck_count_mat, meancut = 0, varcut = 0, title = "time-resolved dcpm")
```

# Integrate Time-resolved RNA TPM for intestine specific genes and ELT-2 single occupancy clusters


```{r}
chip_rna_df <-  as.data.frame(boeck_count_mat) %>% rownames_to_column(var = "wbps_gene_id") %>%
  inner_join(intestine_gene_list,
             by = c("wbps_gene_id" = "WBGeneID")) %>%
  left_join(single_cluster_genes, by = c("wbps_gene_id" = "WBGeneID")) %>%
  dplyr::rename(WBGeneID = wbps_gene_id, cluster.description = single_cluster) %>%
  filter(!is.na(cluster.description))

chip_rna_df$cluster.description <- factor(chip_rna_df$cluster.description, levels = c("Embryo_Specific", "Increasing", "L3_High", "Larval", "Not_Changing"))

nrow(chip_rna_df)
table(chip_rna_df$cluster.description)

# Choose the time-resolve RNA-seq samples
chip_rna_matrix <- chip_rna_df %>% dplyr::select(emb_510min:YA) %>% as.matrix()
rownames(chip_rna_matrix) <- chip_rna_df$WBGeneID
head(chip_rna_matrix)
var_mean_plot(chip_rna_matrix, meancut = 0, varcut = 0, title = "time-resolved intestine gene dcpm")
```

# Are there intestine specific genes that are not ELT-2 bound?

```{r}
intestine_elt2 <- intestine_gene_list %>% full_join(elt2_gene_occupancy, by = "WBGeneID")
intestine_elt2$single_cluster[is.na(intestine_elt2$single_cluster)] <- "Not_Bound"

intestine_elt2 %>% group_by(single_cluster) %>% summarise(gene_total = 100*n()/nrow(intestine_elt2)) %>% ggplot(aes(x = single_cluster, y = gene_total)) +
  geom_bar(stat = "identity")
table(intestine_elt2$single_cluster)
nrow(chip_rna_df)
nrow(intestine_gene_list)
```

# Figure S5A: How many intestine genes are bound by ELT-2 at any stage regardless of occupancy cluster?

```{r}
intestine_gene_list
intestine_peaks <- elt2_peaks %>% group_by(WBGeneID) %>% summarise(peaks = n()) %>% right_join(intestine_gene_list, by = "WBGeneID") %>% dplyr::select(WBGeneID, peaks) %>% replace_na(list(peaks = 0)) %>% mutate(status = case_when(peaks == 0 ~ "not_bound", peaks >= 1 ~ "bound"))
intestine_hyper <- intestine_peaks %>% group_by(status) %>% summarise(genes = n()) 
intestine_hyper
nrow(intestine_gene_list) == sum(intestine_hyper$genes)


my_hyper<- function(x,k,m,n){
  # x = total number of successes in the sample size
  # k = sample size
  # m = population successes
  # n = population non-successes
  dhyper(x = x, k = k, m = m, n = n)
}

hyper_expected <- function(x,k,m,n){
  # x = total number of successes in the sample size
  # k = sample size
  # m = population successes
  # n = population non-successes
  k*m/(m+n)
}

bound_intestine_genes <-  intestine_hyper %>% filter(status == "bound") %>% dplyr::select(genes) %>% as.numeric() # the number of successes in the sample size, the number of bound genes within one set
all_intestine_genes <- sum(intestine_hyper$genes) # sample size, the total number of genes in a single set
all_elt2_bound_genes <- length(unique(elt2_peaks$WBGeneID)) # population successes, number of bound genes in the genome
all_elt2_notbound_genes <- 20470 - all_elt2_bound_genes # number of genes not bound in the genome
# n <- intestine_hyper %>% filter(status == "not_bound") %>% select(genes) %>% as.numeric() # population non-successes, number of not bound genes

my_hyper(x =bound_intestine_genes, k = all_intestine_genes, m = all_elt2_bound_genes, n = all_elt2_notbound_genes)

```


```{r}
phyper(q = bound_intestine_genes-1, m = all_elt2_bound_genes, n = all_elt2_notbound_genes, k = all_intestine_genes, lower.tail = FALSE)
```

```{r fig.width=3, fig.height=3}

num_genes <- 20470
whole_genome_bound <- length(unique(elt2_peaks$WBGeneID))
whole_genome_notbound <- num_genes-whole_genome_bound

whole_and_intestine_binding <- data.frame(
  type = c("whole_genome", "whole_genome", "intestine_genes", "intestine_genes"),
  status = c("elt2_bound", "elt2_notbound", "elt2_bound", "elt2_notbound"),
  genes = c(whole_genome_bound, whole_genome_notbound, bound_intestine_genes, intestine_hyper %>% filter(status == "not_bound") %>% dplyr::select(genes) %>% as.numeric()),
  set_size = c(num_genes, num_genes, all_intestine_genes, all_intestine_genes)
)  %>% mutate(gene_percent = 100*genes/set_size)

whole_and_intestine_percent <- whole_and_intestine_binding %>%
  ggplot(aes(x = type, y = gene_percent, fill = status)) + 
  geom_col(width = 0.75, colour = "black")+
  theme_classic() +
  geom_text(aes(label = paste0(round(gene_percent,1), "%")),
            position = position_stack(vjust = 0.5)) +
  xlab(NULL) +
  ylab("gene %") +
  scale_x_discrete(labels = c("intestine", "whole\ngenome")) +
  labs(fill = "ELT-2 binding") +
  scale_fill_manual(labels = c("bound", "not bound"), values = c("grey", "white"))

whole_and_intestine_percent
```


```{r fig.width=3, fig.height=3}
obs_expect_intestine_bound <- intestine_hyper %>% bind_rows(list(
  status = "bound_expect",
  genes = hyper_expected(
    x = bound_intestine_genes,
    k = all_intestine_genes,
    m = all_elt2_bound_genes,
    n = all_elt2_notbound_genes
  )
)) %>% filter(status != "not_bound") %>% 
  ggplot(aes(x = status, y = genes)) + 
  geom_bar(stat = "identity", width = 0.75, color = "black", fill = "grey") + 
  theme_classic() +
  xlab("ELT-2 bound\nintestine genes") +
  ylab("# of genes") +
  scale_x_discrete(labels = c("observed", "expected"))
obs_expect_intestine_bound
```

save the plots
```{r}
if (plot == TRUE){
ggsave(paste(plotdir,
          "FigureS4_Percent_Genes_ELT2_Bound",
          "_",
          today(),
          ".pdf",
          sep = ""), 
       plot = whole_and_intestine_percent,
       width = 3.25, height = 3, 
       useDingbats=FALSE)
ggsave(paste(plotdir,
          "FigureS4_Expected_Intestine_Genes_ELT2_Bound",
          "_",
          today(),
          ".pdf",
          sep = ""), 
       plot = obs_expect_intestine_bound,
       width = 2, height = 3, 
       useDingbats=FALSE)
}
```


```{r}
whole_and_intestine_binding %>% filter(status == "elt2_bound") %>% mutate(expected = set_size*whole_genome_bound/num_genes)
```
# Figure S5B: Single vs multi cluster
```{r}
head(intestine_elt2)
singlevsmultigenes<- data.frame(genes = c("yes", "yes"),
           gene_type = c("single_occupancy_genes", "multi_occupancy_genes"),
           gene_num = c(sum((intestine_elt2 %>% filter(WBGeneID %in% intestine_gene_list$WBGeneID) %>% group_by(single_cluster) %>% summarise(gene_total = n()) %>% filter(single_cluster %in% elt2_cluster_names))$gene_total),
           (intestine_elt2 %>% filter(WBGeneID %in% intestine_gene_list$WBGeneID) %>% group_by(single_cluster) %>% summarise(gene_total = n()) %>% filter(single_cluster == "Multi_Cluster"))$gene_total))
  
bound_size <- sum(singlevsmultigenes$gene_num)
singlevsmultigenes <- singlevsmultigenes %>% mutate(gene_percent = gene_num*100/bound_size)
singlevsmulti_plot <- singlevsmultigenes %>% 
  ggplot(aes(x = genes, y = gene_percent, fill = gene_type)) +
  geom_col(width = 0.75, colour = "black")+
  theme_classic() +
  geom_text(aes(label = paste0(round(gene_percent,1), "%")),
            position = position_stack(vjust = 0.5)) +
  xlab(NULL) +
  ylab("gene %") +
  scale_x_discrete(labels = c("ELT-2 bound\nintestine genes")) +
  labs(fill = "ELT-2 binding") +
  scale_fill_manual(labels = c("multi_occupancy_genes", "single_occupancy_genes"), values = c("grey", "white"))
singlevsmulti_plot
```

```{r}
single_breakdown_plot <- intestine_elt2 %>% filter(WBGeneID %in% intestine_gene_list$WBGeneID) %>% group_by(single_cluster) %>% summarise(gene_total = n()) %>% filter(single_cluster %in% elt2_cluster_names) %>% mutate(gene_percent = 100*gene_total/sum(gene_total)) %>% 
  ggplot(aes(x = factor(single_cluster, levels = c("Embryo_Specific", "Larval", "Increasing", "L3_High", "Not_Changing"), labels = c("embryo cluster", "larval cluster", "increasing cluster", "L3 cluster", "un-dynamic")), 
             y = gene_percent)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  geom_text(aes(label = paste0(round(gene_percent,1), "%")),
            position = position_dodge(width = 1), vjust = -0.5) +
  xlab("dynamic intestine-specific genes with unique ELT-2 occupancy") +
  ylab("gene %")
single_breakdown_plot
```

save the plots
```{r}
if (plot == TRUE){
ggsave(paste(plotdir,
          "FigureS5B_Percent_Intestine_Genes_ELT2_Occupancy_Type",
          "_",
          today(),
          ".pdf",
          sep = ""), 
       plot = singlevsmulti_plot,
       width = 3.5, height = 3.25, 
       useDingbats=FALSE)
ggsave(paste(plotdir,
          "FigureS5B_Percent_Intestine_Unique_Occupancy_Breakdown",
          "_",
          today(),
          ".pdf",
          sep = ""), 
       plot = single_breakdown_plot,
       width = 5, height = 3.25, 
       useDingbats=FALSE)
}
```

# Variance filter, log transform and z-score row normalize time-resolved RNA expression values


```{r chip_rna_matrix}
#### Handle 0's and take the log
# 1. Just replace 0's as NAs so we can apply log(). Alternatively, we could do log(x + .01), but there are only ~300
chip_rna_matrix_na = chip_rna_matrix;
chip_rna_matrix_na[0 == chip_rna_matrix_na] <- NA
# 2. Apply log()
chip_rna_matrix_log = log( chip_rna_matrix_na )
# 3. Do variances row-wise, make sure to set na.rm=T
rowvariances = apply(chip_rna_matrix, 1, var, na.rm=T)
range(rowvariances) # make sure there are no NaNs
# 4. Exclude the lowest 5% of the rows by their variance
# Remove NA's introduced by log transform
real = apply(chip_rna_matrix_log, 1, function(x) { ! any(is.na(x)) }) 
# Set threshold
thresh = 0.5 #round(ecdf(rowvariances)(chip_rna_matrix_rowvar["WBGene00000007"]), 3)
# Calculate row variance value at threshold
qthreshold = quantile(rowvariances,thresh)
# Identify matrix indexes where variance is above threshold
changing = rowvariances >= qthreshold
# 5. Z score scale the log transformed values
chip_rna_matrix_scaled <- row_scale(chip_rna_matrix_log) # calls base::scale() via RWC23_Functions.R
# chip_rna_matrix_scaled <- t(apply(chip_rna_matrix, 1, function(x)(x-min(x))/(max(x)-min(x)))) # zero to one scaling
```
```{r}
smFISH_genes <- c("aat-6", "pgp-1", "elt-2", "elt-7", "vit-2")
genes_df <- time_resolved_rna %>% filter(wikigene_name %in% smFISH_genes) %>% dplyr::select(wormbase_gseq, wbps_gene_id, wikigene_name)
genes_list <- setNames(genes_df$wikigene_name, genes_df$wbps_gene_id)

chip_rna_matrix_rowmean <- apply(chip_rna_matrix, 1, mean)
chip_rna_matrix_rowvar <- apply(chip_rna_matrix, 1, var)
ix <- rownames(chip_rna_matrix) %in% genes_df$wbps_gene_id

var_mean_plot(chip_rna_matrix, meancut = 0, varcut = qthreshold, sprintf(
    "Time Resolved RNA\nVariance excludes lower %.0f%%",
    thresh * 100))
text(x = log10(chip_rna_matrix_rowmean[ix]),log10(chip_rna_matrix_rowvar[ix]), labels = genes_list[c(rownames(chip_rna_matrix[ix,]))], col = "red", pos = 2, offset = 1.5, cex = 0.8)
points(log2(chip_rna_matrix_rowmean[ix]), log2(chip_rna_matrix_rowvar[ix]), col = "red")


```

# Cluster Time-resolved RNA-seq of intestine genes based on expression patterns

Perform clustering
```{r BoeckRNA_ELT2_chip}
data.whole = chip_rna_matrix_scaled[real&changing,]
set.seed(5)
hc = hclust(dist(data.whole), method ='ward.D2')
```

# Figure S5C: Scree plot analysis of developmental intestine gene expression

```{r BoeckRNA_ELT2_chip}
# plot the hc height to get the screeplot
intestine_scree <- ggplot(NULL, aes(x=length(hc$height):1, y=hc$height)) +
    geom_point() + geom_line() +
    theme_bw() + 
  # labs(title="Scree Plot of Intestine Development Expression", x = "# of clusters", y="Height") + 
  labs(x = "# of clusters", y="dendrogram height") + 
  scale_x_continuous(limits=c(1,15)) +
  theme_classic()
if (plot == TRUE){
ggsave(paste(plotdir,
          "FigureS4_Intestine_Expression_Scree_Plot",
          "_",
          today(),
          ".pdf",
          sep = ""), 
       plot = intestine_scree,
       width = 3, height = 2, 
       useDingbats=FALSE)
}
intestine_scree
```

# Set the number of clusters

```{r BoeckRNA_ELT2_chip}
nclust=4

# clusterboot( dist(data.whole), clustermethod=disthclustCBI,method=agglom,k=nclust)

```
# Collect WBGeneIDs and 

```{r}
smFISH_genes <- c("aat-6", "pgp-1", "elt-2", "elt-7", "vit-2")
genes_df <- time_resolved_rna %>% filter(wikigene_name %in% smFISH_genes) %>% dplyr::select(wormbase_gseq, wbps_gene_id, wikigene_name)
genes_list <- setNames(genes_df$wikigene_name, genes_df$wbps_gene_id)
# genes_list[["C18G1.2"]]
genes_df$wbps_gene_id
genes_list[["WBGene00000007"]]
```


```{r BoeckRNA_ELT2_chip}
#, fig.width = 6, fig.height=4}
# 1. Record matrix indexes for genes of interest to label on heatmap
# gene_names <- c("bcl-11", "sre-23", "T25D10.4", "aat-6", "pgp-1", "ftn-1", "elt-2", "elt-7")
subsetrows=real&changing
ix=which(rownames(chip_rna_matrix_scaled[subsetrows,]) %in% genes_df$wbps_gene_id)
chip_GOI_df = data.frame(name=genes_list[c(rownames(chip_rna_matrix_scaled)[subsetrows][ix])], index=ix)
# 2. Record ELT-2 occupancy pattern annotation per gene
boeck_chip_annotation <- as.data.frame.matrix(table(chip_rna_df$WBGeneID[real&changing], chip_rna_df$cluster.description[real&changing]))
boeck_chip_annotation <- make_cluster_annotation(data.whole, boeck_chip_annotation)
boeck_chip_annotation <- make_cluster_binary_annotation(boeck_chip_annotation)
# 3. Assign cluster names
# rna_cluster_names <- c("One", "Two", "Three", "Four")
rna_cluster_names <- factor(c("One", "Two", "Three", "Four"), levels = c("Two", "Three", "Four", "One"), labels = c("Embryonic", "L1", "Post-embryonic", "Adult"))
# 4. Draw the heatmap
time_resolved_heatmap <- Heatmap(
  data.whole,
  col = colorRampPalette(c("cyan", "black", "yellow"))(1000),
  column_title = sprintf(
    "Boeck Time Resolved RNA\nRow Z-score of Log(RNA Timecourse)\nVariance excludes lower %.0f%% (n = %i)",
    thresh * 100, nrow(data.whole)
  ),
  name = "Row z-score log(dcpm)",
  cluster_columns = F,
  row_order = hc$order,
  row_split = rna_cluster_names[cutree(hc, nclust)],
  row_title = paste(levels(rna_cluster_names), "\n(n = ",as.vector(table(rna_cluster_names[cutree(hc, nclust)])), ")", sep = ""),
  row_title_rot = 0,
  show_row_names = F
) +
  binding_cluster_row_annotation(boeck_chip_annotation)
# +
#   rowAnnotation(foo = anno_mark(at = chip_GOI_df$index,
#                                 labels = chip_GOI_df$name)) 
  # + rowAnnotation(
  #   Not_Bound = boeck_chip_annotation$Not_Bound,
  #   col = list(Not_Bound = c(
  #     "absent" = "white", "present" = "black"
  #   )),
  #   border = TRUE
  # )
time_resolved_heatmap

if (plot == TRUE){
myPDFplot(time_resolved_heatmap, "TimeresolvedRNA_Heatmap_ELT2_annotation", width = 7, height = 6, plotdir)
}

```
Number of genes in each cluster

```{r}
table(rna_cluster_names[cutree(hc, nclust)])
```


Output time-resolved cluster and ELT-2  
```{r}
fig3df <- data.frame(intestine_dev_cluster = rna_cluster_names[cutree(hc, nclust)], WBGeneID = names(cutree(hc, nclust))) %>% dplyr::select(WBGeneID, intestine_dev_cluster)

# write_csv(fig3df, path = "02_output/Fig3_Intestine_Development_Cluster_Assignments.csv")
```
# Make mosaic dot plot

```{r BoeckRNA_ELT2_chip}
mosaic_df <- data.frame(Cluster = rna_cluster_names[cutree(hc, nclust)], WBGeneID = names(cutree(hc, nclust))) %>% left_join(chip_rna_df %>% dplyr::select(WBGeneID, cluster.description), by = "WBGeneID")
mosaic_df$rna_cluster_name <- rna_cluster_names[mosaic_df$Cluster]
mosaic_df$rna_cluster_name <- factor(mosaic_df$rna_cluster_name, levels = rna_cluster_names)
mosaic_df
tab <- table(mosaic_df$Cluster, mosaic_df$cluster.description)
dimnames(tab) <- list(
  "hclusters" = levels(mosaic_df$rna_cluster_name),
  "ELT-2" = levels(chip_rna_df$cluster.description)
)
tab

```


```{r custom-mosaic}
boeck_pearson_residuals <-
  as.data.frame(chisq.test(tab, simulate.p.value = TRUE)$stdres)
boeck_pearson_residuals <-
  cbind(boeck_pearson_residuals,
        as.data.frame(chisq.test(tab, simulate.p.value = TRUE)$observed)$Freq)
colnames(boeck_pearson_residuals) <-
  c("hclusters", "ELT2_cluster", "Pearson_Residual", "Observed")
boeck_pearson_residuals$hclusters <-
  factor(
    boeck_pearson_residuals$hclusters,
    levels = rna_cluster_names[4:1]
  )

# add q-value
q <- p.adjust(
  2*pnorm(
    abs(chisq.test(tab, simulate.p.value = TRUE)$stdres), 
    lower.tail = FALSE), 
  method = "BH")

tableformosiac.q <- matrix(q, ncol = 5)

dimnames(tableformosiac.q) <- dimnames(tab)
tableformosiac.q <- as.table(tableformosiac.q)
boeck_pearson_residuals <-
  cbind(boeck_pearson_residuals, 
        p.adjust = as.data.frame(tableformosiac.q)$Freq)
```


Set factor order and change labels
```{r}
boeck_pearson_residuals$ELT2_cluster <- factor(boeck_pearson_residuals$ELT2_cluster, 
                                               levels = c("Embryo_Specific", 
                                                          "Increasing", 
                                                          "L3_High", 
                                                          "Larval", 
                                                          "Not_Changing"),
                                               labels = c("embryo cluster",
                                                          "larval cluster",
                                                          "increasing cluster",
                                                          "L3 cluster",
                                                          "un-dynamic"))
boeck_pearson_residuals$hclusters <- factor(boeck_pearson_residuals$hclusters, levels = rev(levels(rna_cluster_names)))
```


```{r } 
boeck_mosaic_plot<- ggplot(boeck_pearson_residuals,
       aes(x = ELT2_cluster, y = hclusters)) +
  geom_point(aes(size = Observed, fill = Pearson_Residual, colour = p.adjust < 0.05), shape = 21, stroke = 1) + 
  scale_fill_gradient2(
    high = "red",
    low = "blue",
    mid = "white",
    midpoint = 0,
  ) +
  scale_size_continuous(range = c(1,5), 
                        breaks = c(2,20,200)) +
  scale_color_manual(values = c("light grey", "black")) +  
theme_bw() +
  theme(axis.text.x = element_text(angle = -45, hjust = 0)) +
  # theme(legend.key.width = unit(0.125,"in"), legend.key.height = unit(0.125, "in"), legend.key.size = unit(0.01, "in"), axis.text.x = element_text(angle = -45, hjust = 0), legend.background = element_rect(fill="lightblue", size=0.5, linetype="solid")) +
  theme(legend.key.height = unit(0.125, "in"), legend.margin = margin(0.1,0.1,0.1,0.1)) +
  # guides(fill = )
  guides(fill = guide_colorbar(title = "pearson residual",title.position = "top",label.position = "bottom", direction = "horizontal"),
         size = guide_legend(title = "# genes"),
         color = guide_legend(title = "q < 0.05")) +
  xlab("ELT-2 Binding Cluster\n(from ChIP-seq)") +
  ylab("Time-resolved Gene Cluster\n(from RNA-seq)")
boeck_mosaic_plot
if (plot == TRUE){
ggsave(paste(plotdir,
          "TimeresolvedRNA_ELT2clusters_dotplot",
          "_",
          today(),
          ".pdf",
          sep = ""), 
       plot = boeck_mosaic_plot,
       width = 5, height = 3.75, 
       useDingbats=FALSE)
}


```

# Make Table S4
```{r custom-mosaic}
boeck_contingency <- cbind(boeck_pearson_residuals, p.value = as.data.frame(pnorm(abs(chisq.test(tab, simulate.p.value = TRUE)$stdres), lower.tail = FALSE))$Freq) %>% dplyr::select(Expression.cluster = hclusters, Occupancy.cluster = ELT2_cluster, Genes.observed = Observed, Pearson.residual = Pearson_Residual, p.value, p.adjust)

head(boeck_contingency)
write.csv(boeck_contingency, file = "./02_output/TableS4_chi-square_expression_vs_occupancy.csv", quote = FALSE, row.names = FALSE)
```

# Scratch space

```{r}
# z <- chip_rna_matrix[apply(chip_rna_matrix > 0.07, 1, sum) > 1,]
# z <- chip_rna_matrix_log[real,]
z <- chip_rna_matrix
z_var <- apply(z, 1, var)
z_mean <- apply(z, 1, mean)
plot(log(z_mean), log(z_var), pch='.')
abline(h=log2(1), col='red')
abline(v=log2(1), col='red')

nrow(chip_rna_matrix[which(z_var >= 0.05 & z_mean >= log(0.07)),1:10])
hist(log2(z_var))
hist(log2(z_mean[z_mean >= 0]))
z_filt <- z[which(z_var >= 0.05 & z_mean >= log(0.07)),1:10]
nrow(z_filt)
# scaledata <- t(scale(t(z)))
scale_percent <- t(apply(z_filt, 1, function(x)(x-min(x))/(max(x)-min(x))))
Heatmap(scale_percent)
```
`TPM = A*1/sum(A)*10^6`
`A = total reads mapped to gene * 10^3/gene length in bp`
```{r}
boeck_counts <- read.csv(file = "./01_input/9_Boeck_et_al_2016_time-resolved_transcriptome/wormbase_genes.readcounts_dcpm_per_sample.txt", sep = " ") %>% dplyr::select(WormbaseName, NonRepetitiveLength, ends_with("counts"))

boeck_count_mat <- boeck_counts %>% column_to_rownames(var = "WormbaseName") %>% dplyr::select(ends_with("counts"), -aggregate_polya_counts, -aggregate_totalRNA_counts, -aggregate_counts) %>% as.matrix.data.frame()

```


```{r}
# Length of each gene
gene.length <- boeck_counts$NonRepetitiveLength
# Divide each row by gene length
length_norm <- t(t(boeck_count_mat)/gene.length)

colSums(length_norm, na.rm = TRUE)
t(t(length_norm)*1e6/colSums(t(length_norm)))
colSums(length_norm)

counts_to_tpm <- function(counts, featureLength, meanFragmentLength) {
  
  # Ensure valid arguments.
  stopifnot(length(featureLength) == nrow(counts))
  stopifnot(length(meanFragmentLength) == ncol(counts))
  
  # Compute effective lengths of features in each library.
  effLen <- do.call(cbind, lapply(1:ncol(counts), function(i) {
    featureLength - meanFragmentLength[i] + 1
  }))
  
  # Exclude genes with length less than the mean fragment length.
  idx <- apply(effLen, 1, function(x) min(x) > 1)
  counts <- counts[idx,]
  effLen <- effLen[idx,]
  featureLength <- featureLength[idx]
  
  # Process one column at a time.
  tpm <- do.call(cbind, lapply(1:ncol(counts), function(i) {
    rate = log(counts[,i]) - log(effLen[,i])
    denom = log(sum(exp(rate)))
    exp(rate - denom + log(1e6))
  }))

  # Copy the row and column names from the original matrix.
  colnames(tpm) <- colnames(counts)
  rownames(tpm) <- rownames(counts)
  return(tpm)
}

boeck_tpm <- counts_to_tpm(counts = boeck_count_mat, featureLength = gene.length, meanFragmentLength = rep(1, ncol(boeck_count_mat)))

boeck_dcpm <- read.csv(file = "./01_input/9_Boeck_et_al_2016_time-resolved_transcriptome/wormbase_genes.readcounts_dcpm_per_sample.txt", sep = " ") %>% dplyr::select(WormbaseName, NonRepetitiveLength, ends_with("dcpm"))

boeck_dcpm_mat <- boeck_dcpm %>% column_to_rownames(var = "WormbaseName") %>% dplyr::select(ends_with("dcpm"), -aggregate_polya_dcpm, -aggregate_totalRNA_dcpm, -aggregate_dcpm) %>% as.matrix.data.frame()


z <- boeck_dcpm_mat
z_var <- apply(z, 1, var)
z_mean <- apply(z, 1, mean)
z_sq_stdev <- apply(z, 1, function(x) sqrt(sd(x)))

plot(log2(z_mean), log2(z_var), pch='.')
abline(h=log(1), col='red')
abline(v=log(1), col='red')


```

```{r}
var_mean_plot <- function(mat, meancut, varcut, title = NULL){
 mat_var <- apply(mat, 1, var)
 mat_mean <- apply(mat, 1, mean)
 plot(log2(mat_mean), log2(mat_var), pch='.', main = title)
abline(h=log2(varcut), col='red')
abline(v=log2(meancut), col='red')
}
par(mfrow=c(1,3))
var_mean_plot(mat = boeck_count_mat, varcut = 1, meancut = 1, title = "unified counts")
var_mean_plot(mat = boeck_dcpm_mat*10, varcut = 1, meancut = 1, title = "dcpm")
var_mean_plot(mat = boeck_tpm, varcut = 1, meancut = 1, title = "tpm")

```

See if unified dcpm is similar to TPM

Sample pairs: 
Early embryo, bean: N2_EE_50.210, N2_EE_50.240
Mid embryo, comma/tadpole: N2_EE_50.390, N2_EE_50.420
Late embryo, pretzel: N2_EE_50.690, N2_EE_50.720
L1: "L1_counts", "N2_L1.1_counts"
L2: "L2_counts", "N2_L2.4_counts"
L3: "L3_counts", "N2_L3.1_counts"
L4: "L4_counts", "L4b_counts"
YA: "YA_counts", "N2_Yad.1_counts"

```{r}
# make list of relevant samples
sample_names <- c("N2_EE_50.210", "N2_EE_50.240", 
                         "N2_EE_50.390", "N2_EE_50.420", 
                         "N2_EE_50.690", "N2_EE_50.720", 
                         "L1", "N2_L1.1",
                         "L2", "N2_L2.4",
                         "L3", "N2_L3.1",
                         "L4", "L4b",
                         "YA", "N2_Yad.1")
# give the samples better names
my_names = c("early_emb_1", "early_emb_2", "mid_emb_1", "mid_emb_2", "late_emb_1", "late_emb_2", "L1_1", "L1_2", "L2_1", "L2_2", "L3_1", "L3_2", "L4_1", "L4_2", "YA_1", "YA_2")

# make df that stores to and from column names
boeck_count_samples <- data.frame(count_names = paste(sample_names, "_counts", sep = ""), my_names = my_names)
# extract read counts from file
boeck_raw_counts <- read.csv(file = "./01_input/9_Boeck_et_al_2016_time-resolved_transcriptome/wormbase_genes.readcounts_dcpm_per_sample.txt", sep = " ") %>% dplyr::select(WormbaseName, NonRepetitiveLength, boeck_count_samples$count_names) %>% rename_at(vars(boeck_count_samples$count_names), ~boeck_count_samples$my_names)
# boeck_raw_counts <- read.csv(file = "./01_input/9_Boeck_et_al_2016_time-resolved_transcriptome/wormbase_genes.readcounts_dcpm_per_sample.txt", sep = " ") %>% dplyr::select(WormbaseName, NonRepetitiveLength, contains("count"), -aggregate_polya_counts, -aggregate_totalRNA_counts, -aggregate_counts)
# turn into matrix
boeck_raw_counts_mat <- boeck_raw_counts %>% column_to_rownames(var = "WormbaseName") %>% dplyr::select(-NonRepetitiveLength) %>% as.matrix.data.frame()

# make df that stores to and from column names
boeck_dcpm_samples <- data.frame(count_names = paste(sample_names, "_dcpm", sep = ""), my_names = my_names)
# extract dcpm values and rename
boeck_raw_dcpm <- read.csv(file = "./01_input/9_Boeck_et_al_2016_time-resolved_transcriptome/wormbase_genes.readcounts_dcpm_per_sample.txt", sep = " ") %>% dplyr::select(WormbaseName, NonRepetitiveLength, boeck_dcpm_samples$count_names) %>% rename_at(vars(boeck_dcpm_samples$count_names), ~boeck_dcpm_samples$my_names)
# turn into matrix
boeck_raw_dcpm_mat <- boeck_raw_dcpm %>% column_to_rownames(var = "WormbaseName") %>% dplyr::select(-NonRepetitiveLength) %>% as.matrix.data.frame()
# Calculate TPM
boeck_raw_tpm <- counts_to_tpm(counts = boeck_raw_counts_mat, featureLength = boeck_raw_counts$NonRepetitiveLength, meanFragmentLength = rep(1, ncol(boeck_raw_counts_mat)))
```

```{r}
par(mfrow=c(1,3))
var_mean_plot(mat = boeck_raw_counts_mat, varcut = 1, meancut = 1, title = "raw counts")
var_mean_plot(mat = boeck_raw_dcpm_mat, varcut = 1, meancut = 1, title = "raw dcpm")
var_mean_plot(mat = boeck_raw_tpm, varcut = 1, meancut = 1, title = "raw tpm")
```


```{r}
smFISH_genes <- c("aat-6", "pgp-1", "elt-2", "elt-7", "vit-2")
genes_df <- time_resolved_rna %>% filter(wikigene_name %in% smFISH_genes) %>% dplyr::select(wormbase_gseq, wbps_gene_id, wikigene_name)
genes_list <- setNames(genes_df$wikigene_name, genes_df$wormbase_gseq)
genes_list[["C18G1.2"]]
```


```{r}
var_mean_plot(mat = boeck_raw_tpm, varcut = 1, meancut = 1, title = "raw tpm")

mat <- boeck_raw_tpm[rownames(boeck_raw_tpm) %in% genes_df$wormbase_gseq,]
ix <- rownames(boeck_raw_tpm) %in% genes_df$wormbase_gseq
mat <- boeck_raw_tpm
mat_var <- apply(mat, 1, var)
mat_mean <- apply(mat, 1, mean)
```


```{r fig.width=2, fig.height=2}
mat_mean[ix]
mat_var[ix]
plot(log2(mat_mean), log2(mat_var), pch='.', main = "tpm")
abline(h=log2(2), col='red')
abline(v=log2(2), col='red')
text(x = log2(mat_mean[ix]),log2(mat_var[ix]), labels = genes_list[c(rownames(mat[ix,]))], col = "red", pos = 2, offset = 1.5, cex = 0.8)
points(log2(mat_mean[ix]), log2(mat_var[ix]), col = "red")
```
```{r}
par(mfrow=c(1,2))
hist(boeck_raw_tpm[mat_var >= 1 & mat_mean >= 1,])
hist(log(boeck_raw_tpm[mat_var >= 1 & mat_mean >= 1,]))
```


```{r}
var_mean_plot(boeck_raw_tpm[rownames(boeck_raw_tpm) %in% genes_df$wormbase_gseq,], varcut = 1, meancut = 1, title = "smFISH genes")


```

```{r}
# select intestine expressed genes
boeck_raw_tpm_gut <- matrix_select(boeck_raw_tpm, intestine_gene_list$Sequence.name)
# log transform
min_mean <- 1
min_var <- 1
tpm_gut_mean <- apply(boeck_raw_tpm_gut, 1, mean)
tpm_gut_var <- apply(boeck_raw_tpm_gut, 1, var, na.rm=T)
# tpm_gut_scale <- t(apply(boeck_raw_tpm_gut[tpm_gut_mean >= min_mean & tpm_gut_var >= min_var,], 1, function(x)(x-min(x))/(max(x)-min(x))))
tpm_gut_scale <- row_scale(boeck_raw_tpm_gut[tpm_gut_mean >= min_mean & tpm_gut_var >= min_var,])
hc = hclust(dist(tpm_gut_scale), method =agglom )
# plot the hc height to get the screeplot
ggplot(NULL, aes(x=length(hc$height):1, y=hc$height)) +
    geom_point() + geom_line() +
    theme_bw() + labs(title="Scree Plot of hclust", x = "# of clusters", y="Height") + scale_x_continuous(limits=c(1,15)) 

Heatmap(tpm_gut_scale, cluster_columns = TRUE, show_row_names = FALSE, km = 6)

```



